File: waza_janken.rb

#各モンスターと技が持つタイプ
class Type
	attr_accessor :name
	def do(hp,atk,defence)
	end
end

class Gu < Type
	def initialize()
		@name = "gu"
	end
	def do(hp,atk,defence)
		p "GU"
		return hp ,atk/2 ,defence*2
	end
end

class Cho < Type
	def initialize()
		@name = "cho"
	end
	def do(hp,atk,defence)
		return hp ,atk*2 ,defence/2
	end
end

class Pa < Type
	def initialize()
		@name = "pa"
	end
	def do(hp,atk,defence)
		return hp*4 ,atk/2 ,defence/2
	end
end

#各モンスターが持つ技

class Waza
	attr_accessor :name ,:type
	def initialize(name)
		@name = name
	end

	def do(m,e)
		puts m.name.to_s + "の" + @name.to_s + "!"

		wazatype = self.type.name
		entype = e.type.name
		@state = judge(self.type.name,e.type.name)

	end
end

class Kirisaku < Waza
	def initialize()
		@name = "きりさく"
		@type = Cho.new()
	end
	def do(m,e)
		super(m,e)
		if(@state) then
			p "勝ち"
			puts "攻撃力×2!"
			e.damage(m.atk*2);
		elsif (@state==false) then
			p "負け"
			e.damage(m.atk/2);
		else
			p "あいこ"
		end
	end
end

class Knuckle < Waza
	def initialize()
		@name = "ナックルパンチ"
		@type = Gu.new()
	end
	def do(m,e)
		super(m,e)
		if(@state) then
			p "勝ち"
			puts "攻撃力×2!"
			e.damage(m.atk*2);
		elsif (@state==false) then
			p "負け"
			e.damage(m.atk/2);
		else
			p "あいこ"
		end
	end
end

class Henka < Waza
	def do(m,e)

	end
end


class Monster
	attr_accessor :name ,:hp , :atk ,:def , :type ,:type2 ,:waza1 , :status
	def initialize(name,hp,atk,defence,type)
		@name = name
		@hp = hp
		@def = defence
		@type = type
		@status = ""
	end

	def damage(hp_damage)
		damage = hp_damage-@def
		if(damage>0) then
			puts @name+"に"+damage.to_s + "のダメージ"
			if(@hp<damage)
				puts "死"
				@status = "死"
				@hp = 0
				return false
			end

			@hp -= damage
		else
			puts @name+"に0のダメージ"
		end

		return @hp
	end

	def show_status
		puts "-------"
		puts "name:"+@name
		puts "hp:"+@hp.to_s
		puts "atk:"+@atk.to_s
		puts "def:"+@def.to_s
		puts "type:"+@type.name.to_s
		puts "status:"+@status.to_s
		puts "-------"
	end
end

class Slime < Monster
	def initialize()
		@name = "スライム"
		@hp = 5
		@atk = 5
		@def = 5
		@type = Gu.new()
		@waza1 = Knuckle.new()
		@hp,@atk,@def = self.type.do(@hp,@atk,@def)
	end
end

class SlimeBes < Slime
	def initialize
		super()
		@name = "スライムベス"
		@hp ,@atk,@def= @hp.to_i*2,@atk.to_i*2,@def.to_i*2
		@type2 = Pa.new()
	end
end


class Drakee < Monster
	def initialize()
		@name = "ドラキー"
		@hp = 4
		@atk = 6
		@def =5
		@type = Cho.new()
		@waza1 = Kirisaku.new()
	end
end

class Nyan < Monster
	def initialize
		@name = "プリズニャン"
		@hp = 10
		@atk = 3
		@def = 2
		@type = Pa.new()
	end
end

def judge(j1,j2)
	if(j1==j2)then
		return nil
	elsif(j1=="gu") then
		return j2=="cho"
	elsif(j1=="cho") then
		return j2=="pa"
	elsif(j1=="pa")then
		return j2=="gu"
	end
end

# m = Monster.new("slime",10,10,10,"cho")
s = Slime.new()
d = Drakee.new()
b = SlimeBes.new()
p = Nyan.new()

d.show_status
s.show_status
d.waza1.do(d,s)
s.waza1.do(s,d)
d.show_status
s.show_status

File: waza_janken.cs
using System;

namespace waza_janken
{
	public class State
	{
		public int hp;
		public int atk;
		public int def;
		public State (int hp,int atk,int def)
		{
			this.hp = hp;
			this.atk = atk;
			this.def = def;
		}
	}

	//  各モンスターと技が持つタイプ
	public abstract class Type
	{
		public string name;
		public abstract State _do (int hp, int atk, int defence);
	}

	public class Gu : Type
	{
		public Gu () {this.name = "gu";}
		public override State _do (int hp,int atk,int defence)
		{
			return new State (hp,atk/2,defence*2);
		}
	}
	
	public class Cho : Type
	{
		public Cho () {this.name = "cho";}
		public override State _do (int hp,int atk,int defence)
		{
			return new State(hp,atk*2,defence/2);
		}
	}
	
	public class Pa : Type
	{
		public Pa () {this.name = "pa";}
		public override State _do (int hp,int atk,int defence)
		{
			return new State (hp*4,atk/2,defence/2);
		}
	}

	// 各モンスターが持つ技
	public class Waza
	{
		public string name;
		public Type type;

		public Waza () {}
		public Waza (string name) {this.name = name;}

		public virtual int _do(Monster m,Monster e)
		{
			Game game = new Game ();
			Console.WriteLine (m.name + "の" + this.name + "!");

			string wazatype = this.type.name;
			string entype = e.type.name;
			return game.judge (this.type.name,e.type.name);
		}
	}

	public class Kirisaku : Waza
	{
		public Kirisaku()
		{
			this.name = "きりさく";
			this.type = new Cho ();
		}
		public override int _do (Monster m,Monster e)
		{
			int state = base._do (m,e);
			switch (state)
			{
			case 1:
				Console.WriteLine ("勝ち");
				Console.WriteLine ("攻撃力×2!"); 
				e.damage (m.atk*2);
				break;
			case -1:
				Console.WriteLine ("負け");
				e.damage (m.atk/2);
				break;
			default:
				Console.WriteLine ("あいこ");
				break;
			}
			return state;
		}
	}
	
	public class Knuckle : Waza
	{
		public Knuckle()
		{
			this.name = "ナックルパンチ";
			this.type = new Gu ();
		}
		public override int _do (Monster m,Monster e)
		{
			int state = base._do (m,e);
			switch (state) {
			case 1:
				Console.WriteLine ("勝ち");
				Console.WriteLine ("攻撃力×2!"); 
				e.damage (m.atk*2);
				break;
			case -1:
				Console.WriteLine ("負け");
				e.damage (m.atk/2);
				break;
			default:
				Console.WriteLine ("あいこ");
				break;
			}
			return state;
		}
	}

	public class Henka : Waza
	{
		public Henka ()
		{
		}
		public override int _do (Monster m,Monster e)
		{
			return 0;
		}
	}

	public class Monster
	{
		public string name;
		public int hp;
		public int atk;
		public int def;
		public Type type;
		public Type type2;
		public Waza waza1;
		public string status;

		public int damage (int hp_damage)
		{
			int damage = hp_damage - this.def;
			if (damage > 0) {
				Console.WriteLine (this.name + "に" + damage + "のダメージ");
				if (this.hp < damage) {
					Console.WriteLine ("死");
					this.status = "死";
					this.hp = 0;
					return 0;
				}

				this.hp -= damage;
			} else {
				Console.WriteLine (this.name+"に0のダメージ");
			}

			return this.hp;
		}

		public void show_status ()
		{
			Console.WriteLine ("-------");
			Console.WriteLine ("name:" + this.name);
			Console.WriteLine ("hp:" + this.hp);
			Console.WriteLine ("atk:" + this.atk);
			Console.WriteLine ("def:" + this.def);
			Console.WriteLine ("type:" + this.name);
			Console.WriteLine ("status:" + this.status);
			Console.WriteLine ("-------");
		}
	}

	public class Slime : Monster
	{
		public Slime ()
		{
			this.name = "スライム";
			this.hp = 5;
			this.atk = 5;
			this.def = 5;
			this.type = new Gu ();
			this.waza1 = new Knuckle  ();
		}
	}

	public class SlimeBes : Slime
	{
		public SlimeBes ()
		{
			this.name = "スライムベス";
			this.hp *= 2;
			this.atk *= 2;
			this.def *= 2;
			this.type2 = new Pa ();
		}
	}

	public class Drakee : Monster
	{
		public Drakee ()
		{
			this.name = "ドラキー";
			this.hp = 4;
			this.atk = 6;
			this.def = 5;
			this.type = new Cho ();
			this.waza1 = new Kirisaku ();
		}
	}
	
	public class Nyan : Monster
	{
		public Nyan ()
		{
			this.name = "プリズニャン";
			this.hp = 10;
			this.atk = 3;
			this.def = 2;
			this.type = new Pa ();
		}
	}

	public class Game
	{
		public int judge (string j1,string j2)
		{
			if(j1==j2){
				return 0;
			}else if(j1=="gu"){
				return j2=="cho" ? 1: -1;
			}else if(j1=="cho"){
				return j2=="pa" ? 1 : -1;
			}else if(j1=="pa"){
				return j2=="gu" ? 1 : -1;
			}
			return 0;
		}
	}

	public class MainClass
	{
		public static void Main (string[] argv)
		{
			Monster s = new Slime();
			Monster d = new Drakee();
			Monster b = new SlimeBes();
			Nyan p = new Nyan();
			d.show_status();
			s.show_status();
			d.waza1._do(d,s);
			s.waza1._do(s,d);
			d.show_status();
			s.show_status();
		}
	}
}

